### PR TITLE
Bootstrap improvements (pigpio, tjbotlib, hw tests, run conversation)

### PR DESCRIPTION
**Pull request by bgbraga**
Bootstrap Improvements:

- Move back to nodejs 6.x that is large used and tested version
- Install pigpio
- Install conversation and also tjbotlib
- Try to run conversation if config.js exists or show some guidance

### CODE REVIEW

File: bootstrap/bootstrap.sh [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #!/bin/sh
+TJBOT_FOLDER=$(dirname "$PWD")
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: at some point, we should decide on a fixed location for this

File: bootstrap/bootstrap.sh [hunk 2]
```diff
@@ -28,8 +29,32 @@ sudo apt-get update
 sudo apt-get -y dist-upgrade
 
 #----nodejs install
-curl -sL https://deb.nodesource.com/setup_7.x | sudo -E bash -
+curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
 sudo apt-get install -y nodejs
 
 #----alsa install
 sudo apt-get install -y alsa-base alsa-utils
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: we also need `libasound2-dev` here

File: bootstrap/bootstrap.sh [hunk 3]
```diff
@@ -28,8 +29,32 @@ sudo apt-get update
...
+
+#----install pigpio (npm install pigpio will be exec by tjbotlib package.json)
+curl abyz.co.uk/rpi/pigpio/pigpio.zip > pigpio.zip
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: how come we need to install `pigpio` directly? i thought this gets installed via `npm install` on `tjbotlib`

File: bootstrap/bootstrap.sh [hunk 4]
```diff
@@ -28,8 +29,32 @@ sudo apt-get update
...
+sudo make install
+
+#----install conversation (to install it will resolve tjbotlib and other dependencies)
+cd $TJBOT_FOLDER
+cd recipes/conversation
+
+npm install
+
+#----try to run tjbot
+if [ ! -f "${TJBOT_FOLDER}/recipes/conversation/config.js" ]; then
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: this should probably run through the hardware tests in `bootstrap/tests` instead of running the recipe because this recipe needs watson credentials, which the user is highly likely not to have yet  :)

File: bootstrap/bootstrap.sh (hunk 2 above)
Comment (jweisz) [inline bootstrap/bootstrap.sh]: also, i dont seem to be able to comment on it, but above (lines 11-25), i would make those 3 sections optional and only do them via command line flags, because they're not really necessary to making a functioning tjbot.  we don't need to do any of those on our bots.

- --disable-ipv6
- --enable-google-dns
- --locale=en_US

or something like that.

File: bootstrap/bootstrap.sh [hunk 5]
```diff
@@ -28,8 +29,32 @@ sudo apt-get update
 sudo apt-get -y dist-upgrade
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: does the pi need to be rebooted at this point?  or is it safe to do the rest of the script first, and then reboot the pi?

File: bootstrap/bootstrap.sh (hunk 1 above)
Comment (jweisz) [inline bootstrap/bootstrap.sh]: also, i think we need some more logic here to do a `git clone` if they haven't done it already.. e.g. ideally, we'd like people to be able to bootstrap a tjbot the same way they install node:

```
curl -sL https://raw.githubusercontent.com/ibmtjbot/tjbot/master/bootstrap/bootstrap.sh | sudo -E bash -
```

which means the script probably needs to test if it exists in the filesystem inside a checked out `tjbot` folder, or if not, it needs to `git clone https://github.com/ibmtjbot/tjbot` first.

File: bootstrap/bootstrap.sh (hunk 3 above)
Comment (victordibia) [inline bootstrap/bootstrap.sh]: It has been my experience on some pi's that pigpio did not get installed for some reason ...
and installing pigpio fixed the issue. Btw I used sudo apt-get install pigpio

Comment (jweisz) [inline bootstrap/bootstrap.sh]: oh right right right, I've seen this too -- need to `sudo apt-get install pigpio` before `npm install pigpio` will work, because otherwise it `node-gyp` fail while compiling the native module when the pigpio headers are missing

i recommend just adding `pigpio` to the `sudo apt-get install -y` line above (line 36?)

File: bootstrap/bootstrap.sh (hunk 1 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: Ok, I will do that

File: bootstrap/bootstrap.sh (hunk 2 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: Locale is only to avoid warnings.
IPV6 and extra DNS depends on the internet settings. If IPV4 and IPV6 are both enabled (it is by default) you could have intermittent problems. Sametimes the connection go over IPV4 and sometime using IPV6. It is happening with my rpi. I think only one (IPV4 or IPV6) should be enabled to make sure everything will works. I don't know if a user will detect the intermittent problem and  use the parameter. But I could change that if you prefer.

File: bootstrap/bootstrap.sh (hunk 3 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: hmmm.... I'm following the official installation steps:
https://www.npmjs.com/package/pigpio

I can test with `sudo apt-get install -y` to see if it will works.

File: bootstrap/bootstrap.sh (hunk 4 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: hmmm... that file config.js doesn't exists by default. What I think is: if user already created it they already write the watson credentials. If not bootstrap will show the next steps asking to inform watson credentials.

I didn't see /tests yet. I will take a look.

File: bootstrap/bootstrap.sh (hunk 3 above)
Comment (fivdi) [inline bootstrap/bootstrap.sh]: > It has been my experience on some pi's that pigpio did not get installed for some reason ...

To the best of my knowledge Raspbian Jessie 2016-05-10 or newer comes with pigpio pre-installed. Earlier version of Raspbian don't come with pigpio pre-installed. Raspbian Jessie Lite doesn't come with pigpio pre-installed either.

Were you by any chance using Raspbian Jessie sometimes and Raspbian Jessie Lite other times?

Comment (bgbraga) [inline bootstrap/bootstrap.sh]: I also have problems with missing pigpio. Maybe the problem is that I'm using Raspbian Jessie Lite. I will not use graphical user interface so I think Lite is better option to me.

**Comment by bgbraga** [comment]
Using git clone to get tjbot code I was having an annoying problem with npm install. A lot of deprecated messages and issues about fibers, node-gyp and pigpio.

I think it is related with an old node-gyp version and also some permission issues.
I found a solution that use --unsafe-perm -g node-gyp... now it is working.

sudo npm install --unsafe-perm -g node-gyp

I think I fix almost part of comments/suggestions.

File: README.md [hunk 6]
```diff
@@ -38,6 +38,12 @@ For reference, here is the wiring diagram to hook up the LED and servo to your R
 > For the servo, note that the red (middle) wire is 5v, the brown wire is ground, and the orange wire is data.
 
 # Bring TJBot to Life
+First, make sure you have configured your Raspberry Pi for TJBot.
+If you already have TJBot files:
+    $ cd tjbot/bootstrap && sudo sh bootstrap.sh
```
Comment (jweisz) [inline README.md]: I'd remove this and just assume they haven't done a `git clone` yet.

File: README.md [hunk 7]
```diff
@@ -38,6 +38,12 @@ For reference, here is the wiring diagram to hook up the LED and servo to your R
...
+If it is your first time, just run that command to download and install TJBot:
+    $ curl -sL https://raw.githubusercontent.com/ibmtjbot/tjbot/master/bootstrap/bootstrap.sh | sudo -E bash -
```
Comment (jweisz) [inline README.md]: should we `ibm.biz`-ify this url?  most people probably won't care, some people would consider it super dangerous to run a bash script from a URL redirector.  :D

File: bootstrap/README.md [hunk 8]
```diff
@@ -50,9 +56,9 @@ If you have plugged in your speaker via the headphone jack, you may experience i
 
 8. Run a recipe
 
-    cd ~/Desktop/tjbot/recipes/intro
```
Comment (jweisz) [inline bootstrap/README.md]: maybe we should make the `intro` script part of bootstrap?  @maryamashoori didn't want to include it as a mainline recipe, but I feel we should have a script that runs and indicates some kind of success that bootstrap worked.

File: bootstrap/bootstrap.sh [hunk 9]
```diff
@@ -28,8 +29,54 @@ sudo apt-get update
...
-sudo apt-get install -y alsa-base alsa-utils
+#----install official requirements
+sudo apt-get install -y alsa-base alsa-utils libasound2-dev
+
+#----install missing pigpio in Raspbian Lite (the command npm install pigpio will be exec by package.json)
+sudo apt-get install pigpio
+
+#----install git and download tjbot
+if [ ! -d "${TJBOT_FOLDER}/recipes/conversation" ]; then
+    sudo apt-get install -y git
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: maybe i'm not following the full logic here, but you're assuming that git wasn't installed if the `TJBOT_FOLDER` is missing?  shouldn't we just include git along with the alsa stuff if there's a chance it doesn't come installed on a stock pi?

File: bootstrap/bootstrap.sh [hunk 10]
```diff
@@ -28,8 +29,54 @@ sudo apt-get update
...
+
+    rm -Rf /home/pi/git/tjbot
+    git clone https://github.com/ibmtjbot/tjbot.git /home/pi/git/tjbot
+
+    TJBOT_FOLDER='/home/pi/git/tjbot'
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: why not `/home/pi/tjbot` or `/home/pi/Desktop/tjbot` or `/home/pi/Documents/tjbot` ?  :)

File: bootstrap/bootstrap.sh [hunk 11]
```diff
@@ -28,8 +29,54 @@ sudo apt-get update
...
+echo "------------------------------------";
+echo "INSTALLATION COMPLETED!!! ;)"
+echo "------------------------------------";
+
+#----test hardware
+cd $TJBOT_FOLDER
+cd bootstrap/tests
+echo "path: $PWD"
+
+sudo npm install --unsafe-perm -g readline-sync
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: we do this twice?

File: bootstrap/bootstrap.sh [hunk 12]
```diff
@@ -28,8 +29,54 @@ sudo apt-get update
...
+
+    TJBOT_FOLDER='/home/pi/git/tjbot'
+fi
+
+#----install conversation (to install it will resolve tjbotlib and other dependencies)
+cd $TJBOT_FOLDER
+cd recipes/conversation
+echo "path: $PWD"
+
+sudo npm install --unsafe-perm -g node-gyp
```
Comment (jweisz) [inline bootstrap/bootstrap.sh]: never needed to do this before...

File: bootstrap/tests/package.json [hunk 13]
```diff
@@ -2,13 +2,13 @@
   "name": "tests",
   "description": "TJBot hardware tests",
   "version": "1.0.0",
-  "author": "Justin Weisz <jweisz@us.ibm.com>",
+  "author": "TJBot Team",
```
Comment (jweisz) [inline bootstrap/tests/package.json]: If we're moving away from individual attribution, which is fine, then I recommend changing this to `TJBot <tjbot@us.ibm.com>`

File: README.md (hunk 6 above)
Comment (bgbraga) [inline README.md]: ok

File: bootstrap/bootstrap.sh (hunk 9 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: The logic is:
If recipes doesn't exists, TJBot is not downloaded yet (scenario that user are running bootstrap.sh from URL).
So that code is getting TJ using git clone.

I'm assuming that if rpi already have TJBot files the user don't need to install git.
I don't know if "already have TJBot files" is a valid scenario. If not, sure, we could move that apt-get to run together alsa.

File: bootstrap/bootstrap.sh (hunk 10 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: kkkkk, it is because I have to test script with two tjbot: one from remote bootstrap.sh and other local version.
I will use /home/pi/tjbot it is better I think.

File: bootstrap/bootstrap.sh (hunk 12 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: yeah... it is strange... if you could test using only sudo npm install ...
here it doesn't works.. I comment on revision - I got a lot of issues with permission, old node-gyp version and other things. I found that command in some node.js issue and it solved all the problems...

sudo npm install --unsafe-perm -g node-gyp

File: bootstrap/bootstrap.sh (hunk 11 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: It is to install test and another to install conversation.
I already have problem to run without --unsafe-perm -g here... If you could test with you rpi...

File: bootstrap/tests/package.json (hunk 13 above)
Comment (bgbraga) [inline bootstrap/tests/package.json]: done

File: bootstrap/bootstrap.sh (hunk 5 above)
Comment (victordibia) [inline bootstrap/bootstrap.sh]: No reboot needed from my experience.

File: bootstrap/README.md (hunk 8 above)
Comment (victordibia) [inline bootstrap/README.md]: Im not sure how good an idea this is.
The `intro recipe script` (and any other recipe) need credentials to work. If boostrap is run to get things installed as a first step, it is unlikely there will be credentials to setup to run. I feel running the set of tests serve as a good way to test the pi hardware setup. Intro might be testing things beyond this i.e. connection to Watson services.

File: bootstrap/bootstrap.sh (hunk 9 above)
Comment (victordibia) [inline bootstrap/bootstrap.sh]: We can move to run with alsa. If its already installed it will not be reinstalled.

File: bootstrap/bootstrap.sh (hunk 12 above)
Comment (victordibia) [inline bootstrap/bootstrap.sh]: I have seen had similar situations where I needed` --unsafe-perm` to proceed. I cant remember what was the issue, but I think it was again related to npm install for pigpio.

File: MAINTAINERS.md [hunk 14]
```diff
@@ -5,3 +5,4 @@ TJBot was lovingly created by researchers at IBM Research.
 - [Maryam Ashoori](https://github.com/maryamashoori) is the creator of TJBot and is the ringleader of our merry band of cardboard robots.
 - [Justin Weisz](https://github.com/jweisz) created the TJBot library and generally tries to bring order to our chaos.
 - [Victor Dibia](https://github.com/victordibia) created TJBot’s initial receipes and is our master of demos.
+- [Bruno Braga](https://github.com/bgbraga) created TJBot bootstrap and helping to maintain the project.
```
Comment (victordibia) [inline MAINTAINERS.md]: Im not sure the entire content of this document is the right way to report attribution. 
I would leave it unedited and subject for a different PR when we have had a chance to discuss this on slack.

Comment (bgbraga) [inline MAINTAINERS.md]: I think it could have only github link and names. Similar with Node.js:
https://github.com/nodejs/node#collaborators

But I removed my change :P

File: bootstrap/bootstrap.sh (hunk 9 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: done

File: bootstrap/README.md (hunk 8 above)
Comment (jweisz) [inline bootstrap/README.md]: we could make intro not require watson credentails, but that's fine.. we can just run the HW tests.

File: bootstrap/bootstrap.sh (hunk 5 above)
Comment (jweisz) [inline bootstrap/bootstrap.sh]: is a reboot needed after the bootstrap finishes running?

File: bootstrap/bootstrap.sh (hunk 11 above)
Comment (jweisz) [inline bootstrap/bootstrap.sh]: ah i misread this, i saw the `--unsafe-perm` and assumed you were installing the same package.  my bad  ;)

File: MAINTAINERS.md (hunk 14 above)
Comment (victordibia) [inline MAINTAINERS.md]: I tend to agree with this approach. Its what we had until recently. Will work on a PR for this to possibly reflect a core team and collaborators as the project grows.

File: bootstrap/bootstrap.sh (hunk 5 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: No. It is not necessary. After bootstrap everything is fine. Except one thing: IPV4 and IPV6 will continue active and after reboot all connections will be using IPV4.

File: bootstrap/bootstrap.sh (hunk 12 above)
Comment (fivdi) [inline bootstrap/bootstrap.sh]: I have never had to do `sudo npm install --unsafe-perm -g node-gyp` before either. In addition, this command simply installs the `node-gyp` package globally in `/usr/lib/node_modules`. It doesn't "install conversation" in `/home/pi/tjbot/recipes/conversation` which according to the comment on line 49 is the intended purpose. The command `npm install` (WITHOUT sudo and WITHOUT --unsafe-perm) appears to be missing.

> I have seen had similar situations where I needed--unsafe-perm to proceed. I cant remember what was the issue, but I think it was again related to npm install for pigpio.

If this was the case then something else was messed up beforehand. Calling `npm install` in `/home/pi/tjbot/recipes/conversation` should be enough to "install conversation".

File: bootstrap/README.md [hunk 15]
```diff
@@ -61,8 +67,8 @@ Tests can be run in the following manner.
...
 ```
 $ npm install
-$ sudo node test/test.camera.js
-$ sudo node test/test.led.js
-$ sudo node test/test.servo.js
-$ sudo node test/test.speaker.js
+$ sudo node tests/test.camera.js
+$ sudo node tests/test.led.js
+$ sudo node tests/test.servo.js
+$ sudo node tests/test.speaker.js
```
Comment (fivdi) [inline bootstrap/README.md]: Removing the $ at the start of the above lines would make cut and paste much easier. To see this, go [here](https://github.com/ibmtjbot/tjbot/blob/feature/bb-bootstrap-initial/bootstrap/README.md#hardware-tests) and try to copy the five lines and paste them into terminal. Or double click a single line, copy it, and paste it to a terminal. The $ is also copied which is painful.

Removing the $ would also be more consistent with most of the other command shown in the README.

File: README.md [hunk 16]
```diff
@@ -38,6 +38,10 @@ For reference, here is the wiring diagram to hook up the LED and servo to your R
 > For the servo, note that the red (middle) wire is 5v, the brown wire is ground, and the orange wire is data.
 
 # Bring TJBot to Life
+First, make sure you have configured your Raspberry Pi for TJBot.
+Just run that command to download and install TJBot:
+    $ curl -sL https://raw.githubusercontent.com/ibmtjbot/tjbot/master/bootstrap/bootstrap.sh | sudo -E bash -
```
Comment (fivdi) [inline README.md]: Removing the $ at the start of the line would make double clicking the line for a cut and paste operation easier. Try it [here](https://github.com/ibmtjbot/tjbot/blob/feature/bb-bootstrap-initial/bootstrap/README.md#tjbot-bootstrap)

Removing the $ would also be more consistent with other command in the README.

File: bootstrap/bootstrap.sh (hunk 12 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: hmmm, I understand about node-gyp...
But could anyone get that bootstrap.sh, change to `npm install` or `sudo npm install` and test in your rpi?
It doesn't not works here.

Comment (fivdi) [inline bootstrap/bootstrap.sh]: @bgbraga  It worked for me manually. I'll give it a try with bootstrap.sh. Can you post the complete output of `npm install` from your Raspberry Pi to a gist?

Comment (bgbraga) [inline bootstrap/bootstrap.sh]: @fivdi see that attachment. It is from npm install of tests folder but the error is the same for conversation.
[install tests.txt](https://github.com/ibmtjbot/tjbot/files/931296/install.tests.txt)

I'm just thinking a thing... I used a similar command to avoid these errors in tests folder:
sudo npm install --unsafe-perm -g readline-sync

It really install tests because the tests are running fine from bootstrap.sh. So now I'm a little confused with you comment: 
> this command simply installs the node-gyp package globally in /usr/lib/node_modules. It doesn't "install conversation"

If `-g readline-sync` is just installing readline-sync, why tests are running?

Comment (fivdi) [inline bootstrap/bootstrap.sh]: The target is moving :). Up till now I thought we were discussing what happens in `/home/pi/tjbot/recipes/conversation`, but now we're talking about `/home/pi/tjbot/bootstrap/tests`.

Are you 100% sure you were logged on as user pi and that the file attached above is the output of `npm install` and NOT the output of `sudo npm install`? Because it looks like the output of `sudo npm install` which is a completely different thing.

What I see when `npm install` is called in `/home/pi/tjbot/bootstrap/tests` can be seen at https://gist.github.com/fivdi/f7a45e595d3039cb0c5d631425bb080d

At the end if the gist `sudo node test.servo.js` is successfully called.

Comment (bgbraga) [inline bootstrap/bootstrap.sh]: @fivdi it is because I'm in an IBM traveling and don't have rpi here. So I get log from my last try that is related with tests folder. But at conversation the behavior is the same: permission errors (even using sudo) that could be fixed using `--unsafe-perm` and also a lot of warnings and deprecated messages. I used "sudo npm install" in that last try. That is true.

Yeah, I know at the end test.servo.js is successfully called. But that a lot of warning / deprecated messages (https://gist.github.com/fivdi/f7a45e595d3039cb0c5d631425bb080d) give the impression that are something wrong. A end user could thing that it is broken.

So `sudo npm install --unsafe-perm -g readline-sync` run tests fine without errors and also `sudo npm install --unsafe-perm -g node-gyp ` do the same with conversation.

I tried to know what is happening with `sudo npm install` or `npm install` but I didn't know. If we have a clean messages using `npm install` sure it is better.

Comment (fivdi) [inline bootstrap/bootstrap.sh]: > permission errors (even using sudo) that could be fixed using --unsafe-perm
> I tried to know what is happening with sudo npm install or npm install but I didn't know. 

It's not easy to understand what npm does when called with sudo when you are installing native modules that need to be compiled. I'll attempt to explain it.

If you cd to the directory of a native module that needs to be compiled and call the command `sudo npm install` in that directory npm will NOT execute everything with root privileges. In particular, when running the script to compile the native module, the process that does the compile has the privileges of user `nobody`. This is not a lot of privileges and explains why errors such as the following are displayed:

```
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/6.10.2"
```

npm does this as running scripts is security risk. The workaround is to run the command `sudo npm install --unsafe-perm` instead. This gives the process that does the compile root privileges.

The npm documentation mentions the topic [here](https://docs.npmjs.com/misc/config#unsafe-perm), [here](https://docs.npmjs.com/misc/config#user) and [here](https://docs.npmjs.com/misc/scripts#user).

That being said, I don't see any reason why `sudo npm install --unsafe-perm` needs to be called here. `npm install` should be enough. If it's not enough something is messed up on your system.

> and also a lot of warnings and deprecated messages
> But that a lot of warning / deprecated messages (https://gist.github.com/fivdi/f7a45e595d3039cb0c5d631425bb080d) give the impression that are something wrong. A end user could thing that it is broken.

Yes, this is correct. Several native modules are being installed and lots of them result in warnings being displayed. But they are only warnings, not errors. I hope you noticed that `pigpio` compiles cleanly and doesn't display any warnings :). I don't think you'll be able to do much about this other than document that fact that the warnings are displayed so that users feel more at ease.

> So sudo npm install --unsafe-perm -g readline-sync run tests fine without errors

I don't understand this. Haw can running the command `sudo npm install --unsafe-perm -g readline-sync` run tests?

> and also sudo npm install --unsafe-perm -g node-gyp do the same with conversation.

I don't understand this either. How can running the command `sudo npm install --unsafe-perm -g node-gyp` run tests?

Comment (bgbraga) [inline bootstrap/bootstrap.sh]: I guess they install everything from package.json as global.

From doc:
> In global mode (ie, with -g or --global appended to the command), it installs the current package context (ie, the current working directory) as a global package.
https://docs.npmjs.com/cli/install

We could do more tests without sudo and not global.
Have you tried to used `npm install` in bootstrap.sh using your rpi?

Comment (fivdi) [inline bootstrap/bootstrap.sh]: I'm not near a Pi at the moment and can't perform any tests. However, if I'm not mistaken, I think the statement from npm above only applies when in a package directory and no package is specified on the command line. In our case here the packages `readline-sync` or `node-gyp` are specified on the command line.

My current understanding is that calling `sudo npm install --unsafe-perm -g readline-sync` in directory `/home/pi/tjbot/bootstrap/tests` will install package `readline-sync` globally in `/usr/lib/node_modules`. It will not install the current package context (`/home/pi/tjbot/bootstrap/tests`) globally.

Calling `sudo npm install --unsafe-perm -g` in directory `home/pi/tjbot/bootstrap/tests` will install the current package context (`/home/pi/tjbot/bootstrap/tests`) globally.

But as already mentioned I'm not near a Pi to try it out.

> We could do more tests without sudo and not global.

This is a good idea. To be honest, I think it's essential.

> Have you tried to used `npm install` in bootstrap.sh using your rpi?

Yes, I tried it yesterday but only for `home/pi/tjbot/recipes/conversation`. I'll try the whole thing again when I'm back at the Pi.

I was just wondering, is it possible that you called `sudo npm install --unsafe-perm -g` in directory `home/pi/tjbot/bootstrap/tests` or `/home/pi/tjbot/recipes/conversation` at some point in the past? If this is the case, the corresponding packages will have been installed globally in the past and have been there ever since. This would explain how tests are still working now.

Note that `bootstrap.sh` only partially cleans up an old installation. It removes local files with `rm -Rf /home/pi/tjbot` but it doesn't uninstall npm packages that are installed globally. This is another reason not to install globally.

And another thought, if users edit/create files in `/home/pi/tjbot` and call `bootstrap.sh` a second time, they may not be too happy that the call to `rm -Rf /home/pi/tjbot` will delete their work.

Comment (fivdi) [inline bootstrap/bootstrap.sh]: I just ran bootstrap.sh on a freshly installed Raspbian Jessie Lite 2017-04-10. bootstrap.sh was modified to use `npm install` rather than `sudo npm install --unsafe-perm -g node-gyp` and `sudo npm install --unsafe-perm -g readline-sync`.

There's a gist [here](https://gist.github.com/fivdi/4be4f73ac08b9e42d74b71e893db96f0) with all the relevant information:
 * `bootstrap.log` - The output of the installation script
 * `bootstrap.sh` - The modified installation script
 * `npm list` - The output of `npm list` after the installation when called in `~/tjbot/bootstrap/tests` to verify that everything was installed locally
 * `npm list -g` - The output of `npm list -g` after the installation when called in `~/tjbot/bootstrap/tests` to ensure that nothing was installed globally

Note that I don't actually have the appropriate hardware attached to the Pi. The test software doesn't know this most of the time but it does with the camera so there was an error with the camera test. There was also an error with the servo test because I was hitting the keyboard too fast and answered one of the questions before it was asked.

I also verified that calling `sudo npm install --unsafe-perm -g readline-sync` in directory `/home/pi/tjbot/bootstrap/tests` will only install package `readline-sync` globally in `/usr/lib/node_modules` and will not install the current package context (/home/pi/tjbot/bootstrap/tests) globally. This probably means that you installed some of the required packages globally in the past and they are still being used today.

File: bootstrap/bootstrap.sh (hunk 11 above)
Comment (jelcaf) [inline bootstrap/bootstrap.sh]: I think that it's necessary `npm install` in `tests` directory (also in `recipes/conversation`).

File: bootstrap/bootstrap.sh (hunk 12 above)
Comment (bgbraga) [inline bootstrap/bootstrap.sh]: @fivdi I also just format my SD card and installed Raspbian Jessie Lite 2017-04-10 to test in a clean rpi the command `npm install` without sudo and other params.
Result: 
[bootstrap.txt](https://github.com/ibmtjbot/tjbot/files/949700/bootstrap.txt)

It is similar with your results but I have an extra issue:
```
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/6.10.2"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/home/pi/tjbot/recipes/conversation/node_modules/fibers/.node-gyp"
```

In general it is solved with `--unsafe-perm` but I didn't used it to keep in the same level of you.

As we can see in our logs, if we use only `npm install` we receive a lot of warnings in installation process. I don't know if it is related with a wrong package version or other thing.
Do you know how to solve these warning messages? I know it is only warning but is a lot of then and give the impression of an installation problem.

As we see I only could avoid that using -g parameter. You have a suspect that it install only a specific package as global and not install package.json. For me it install the -g parameter as global **and also** package.json as local because the folder `node_modules` is always created inside the conversation and tests folder.

BTW, do you know any other way to avoid these warning messages?

Comment (fivdi) [inline bootstrap/bootstrap.sh]: @bgbraga 

> ```
> gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/6.10.2"
> gyp WARN EACCES attempting to reinstall using temporary dev dir "/home/pi/tjbot/recipes/conversation/node_modules/fibers/.node-gyp"
> ```
The EACCES warnings are occurring because bootstrap.sh was called with sudo from the command line. Try with `./bootstrap.sh ` rather than `sudo ./bootstrap.sh` and the EACCES warnings should no longer occur.

>As we can see in our logs, if we use only npm install we receive a lot of warnings in installation process. I don't know if it is related with a wrong package version or other thing.
>Do you know how to solve these warning messages? I know it is only warning but is a lot of then and give the impression of an installation problem.

I can't tell you if it's related to the wrong package versions. I assume that the authors tjbot and tjbotlib checked that the versions of the packages being used can be used. It's most likely the case that the authors of the native packages simply haven't updated their packages to prevent the warnings in Node.js v6 yet. I think you have the following options:
 * Live with the warnings and document them well so that users feel more at ease
 * Downgrade to Node.js v4 to see if there are no warnings or less warnings
 * Create issues at the corresponding repositories asking the authors if the packages can be updated and maybe create pull requests if they can be updated
 * Hide the output from the user with command line redirection, for example, something like `npm install > install.log 2>&1` might work.

Comment (bgbraga) [inline bootstrap/bootstrap.sh]: @jweisz @victordibia I tested with npm 7.x to see if it resolve the warning and deprecated messages, but not lucky.

I also see that the packages that are causing these messages are out of control of TJBot. It is some dependencies of dependencies.

So we have to decide what to do here to move on. 
I revert the last commit to use "npm install". 
It is already tested by me and @fivdi, it is not have a beautiful output but works.
To minimize the problem I also silence the output with `npm install > install.log 2>&1` (same suggested by fivdi).

What do you think is better?
1) keep `npm install` and document the annoying messages
2) use lastest commit to hide the detail of npm install and have an output like that:

> Reading package lists... Done
> 
> ## Run `apt-get install nodejs` (as root) to install Node.js v6.x and npm
> 
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done
> The following NEW packages will be installed:
>   nodejs
> 0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
> Need to get 0 B/8,850 kB of archives.
> After this operation, 45.4 MB of additional disk space will be used.
> Selecting previously unselected package nodejs.
> (Reading database ... 32391 files and directories currently installed.)
> Preparing to unpack .../nodejs_6.10.2-1nodesource1~jessie1_armhf.deb ...
> Unpacking nodejs (6.10.2-1nodesource1~jessie1) ...
> Processing triggers for man-db (2.7.0.2-5) ...
> Setting up nodejs (6.10.2-1nodesource1~jessie1) ...
> Installing new version of config file /etc/profile.d/nodejs.sh ...
> Reading package lists... Done
> Building dependency tree       
> Reading state information... Done

File: bootstrap/README.md [hunk 17]
```diff
@@ -7,7 +7,7 @@ Perform the following operations to prepare your Raspberry Pi for becoming a TJB
 
 2. Run the command below:
 
-    $ curl -sL https://raw.githubusercontent.com/ibmtjbot/tjbot/master/bootstrap/bootstrap.sh | sudo -E bash -
+    curl -sL https://raw.githubusercontent.com/ibmtjbot/tjbot/master/bootstrap/bootstrap.sh | sudo -E bash -
```
Comment (fivdi) [inline bootstrap/README.md]: Nice to see the $ removed to make cut and paste easier :+1: 

I'm sorry I didn't spot this the last time, but if I'm not mistaken `sudo` shouldn't be used here as it will result in unwanted behavior when `npm install` is called from `bootstrap.sh`. Remember, `npm install` should be called without `sudo`.
